examples/multi-token/test-contract-defi/src/lib.rs

```diff
@@ -0,0 +1,98 @@
+/*!
+Some hypothetical DeFi contract that will do smart things with the transferred tokens
+*/
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::U128;
+use near_sdk::{
+    env, log, near_bindgen, require, AccountId, Balance, Gas, PanicOnDefault,
+    PromiseOrValue,
+};
+use near_contract_standards::multi_token::core::MultiTokenReceiver;
+use near_contract_standards::multi_token::token::TokenId;
+
+const BASE_GAS: u64 = 5_000_000_000_000;
+const PROMISE_CALL: u64 = 5_000_000_000_000;
+const GAS_FOR_MT_ON_TRANSFER: Gas = Gas(BASE_GAS + PROMISE_CALL);
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct DeFi {
+    multi_token_account_id: AccountId,
+}
+
+// Have to repeat the same trait for our own implementation.
+trait ValueReturnTrait {
+    fn value_please(
+        &self,
+        num_tokens: usize,
+        amount_to_return: String,
+    ) -> PromiseOrValue<Vec<U128>>;
+}
+
+#[near_bindgen]
+impl DeFi {
+    #[init]
+    pub fn new(multi_token_account_id: AccountId) -> Self {
+        require!(!env::state_exists(), "Already initialized");
+        Self { multi_token_account_id }
+    }
+}
+
+#[near_bindgen]
+impl MultiTokenReceiver for DeFi {
+    /// If given `msg: "take-my-money", immediately returns U128::From(0)
+    /// Otherwise, makes a cross-contract call to own `value_please` function, passing `msg`
+    /// value_please will attempt to parse `msg` as an integer and return a vec of
+    /// token_ids.len() many copies of the U128 version of it.
+    fn mt_on_transfer(
+        &mut self,
+        sender_id: AccountId,
+        previous_owner_ids: Vec<AccountId>,
+        token_ids: Vec<TokenId>,
+        amounts: Vec<U128>,
+        msg: String,
+    ) -> PromiseOrValue<Vec<U128>> {
```

Comment:
From the [NEP-245](https://nomicon.io/Standards/Tokens/MultiToken/Core) 
```
 Take some action after receiving a multi token

 Requirements:
 * Contract MUST restrict calls to this function to a set of whitelisted  
   contracts
 * Contract MUST panic if `token_ids` length does not equals `amounts`
   length
 * Contract MUST panic if `previous_owner_ids` length does not equals `token_ids`
  length
 ```
But, you provided a nice case, I think It would be nice to add more test cases.

Reply:
from other side returning Vec<(TokenId, U128)> not safe because a predecessor can change the token

Reply:
> from other side returning Vec<(TokenId, U128)> not safe because a predecessor can change the token

What do you mean?

Well, thinking about it a bit more, I feel that the list of amounts could be cheaper/easier to implement, so maybe it is not worth complicating things

Reply:
I kept parameters as-is based on the comments above. Agree with @frol that prioritizing simplicity makes sense here